pkg/spoke/addon/configuration.go

```diff
@@ -46,91 +30,69 @@ type registrationConfig struct {
 
 func (c *registrationConfig) x509Subject(clusterName, agentName string) *pkix.Name {
 	subject := &pkix.Name{
-		CommonName:         c.Subject.CommonName,
-		Organization:       c.Subject.Organization,
-		OrganizationalUnit: c.Subject.OrganizationalUnit,
+		CommonName:         c.registration.Subject.User,
+		Organization:       c.registration.Subject.Groups,
+		OrganizationalUnit: c.registration.Subject.OrganizationUnits,
 	}
 
 	// set the default common name
 	if len(subject.CommonName) == 0 {
-		subject.CommonName = defaultCommonName(clusterName, agentName, c.AddOnName)
+		subject.CommonName = defaultCommonName(clusterName, agentName, c.addOnName)
 	}
 
 	// set the default organization if signer is KubeAPIServerClientSignerName
-	if c.SignerName == certificatesv1.KubeAPIServerClientSignerName && len(subject.Organization) == 0 {
-		subject.Organization = []string{defaultOrganization(clusterName, c.AddOnName)}
+	if c.registration.SignerName == certificatesv1.KubeAPIServerClientSignerName && len(subject.Organization) == 0 {
+		subject.Organization = []string{defaultOrganization(clusterName, c.addOnName)}
 	}
 
 	return subject
 }
 
-// certSubject contains fields which are used to generate csr subject and can be customized by users
-type certSubject struct {
-	CommonName         string   `json:"commonName,omitempty"`
-	Organization       []string `json:"organization,omitempty"`
-	OrganizationalUnit []string `json:"organizationalUnit,omitempty"`
-}
-
-// getAddOnConfig returns addon configuration information from addon annotations.
-func getAddOnConfig(addOn *addonv1alpha1.ManagedClusterAddOn) (*addOnConfig, error) {
-	installationNamespace, ok := addOn.Annotations[installationNamespaceAnnotation]
-	if !ok {
+// getAddOnInstallationNamespace returns addon installation namespace from addon spec.
+// If the installation namespace is not specficed on addon spec, the addon defaul
+// installation namespace open-cluster-management-agent-addon will be returned.
+func getAddOnInstallationNamespace(addOn *addonv1alpha1.ManagedClusterAddOn) string {
+	installationNamespace := addOn.Spec.InstallNamespace
+	if installationNamespace == "" {
 		installationNamespace = defaultAddOnInstallationNamespace
 	}
 
-	addOnConfig := &addOnConfig{
-		AddOnName:             addOn.Name,
-		InstallationNamespace: installationNamespace,
-	}
-
-	return addOnConfig, nil
+	return installationNamespace
 }
 
 // getRegistrationConfigs reads annotations of a addon and returns a map of registrationConfig whose
 // key is the hash of the registrationConfig
-func getRegistrationConfigs(addOnName string, annotations map[string]string) (map[string]registrationConfig, error) {
-	ns := annotations[installationNamespaceAnnotation]
-	if len(ns) == 0 {
+func getRegistrationConfigs(addOn *addonv1alpha1.ManagedClusterAddOn) (map[string]registrationConfig, error) {
+	if len(addOn.Status.Registrations) == 0 {
```

Comment:
nit: I think this is unnecessary, wince without registrations, that loop doesn't run and you return empty.

Reply:
yeah, this is indeed unnecessary, removed